File: pkg/node/ratelimit.go
package node

import (
	"sync"
	"time"

	pb "github.com/ivere27/nitella/pkg/api/proxy"
)

// RateLimiter manages rate limiting and auto-blocking state
type RateLimiter struct {
	config *pb.RateLimitConfig
	mu     sync.Mutex

	// IP -> state
	counters map[string]*ipState
	blocks   map[string]time.Time // IP -> Block expiration time

	// Cleanup control
	quit chan struct{}
}

type ipState struct {
	count       int
	firstSeen   time.Time
	lastSeen    time.Time
	failedCount int
	banLevel    int
}

func NewRateLimiter(config *pb.RateLimitConfig) *RateLimiter {
	if config == nil {
		return nil
	}
	rl := &RateLimiter{
		config:   config,
		counters: make(map[string]*ipState),
		blocks:   make(map[string]time.Time),
		quit:     make(chan struct{}),
	}
	go rl.cleanupLoop()
	return rl
}

// Stop stops the cleanup goroutine
func (rl *RateLimiter) Stop() {
	close(rl.quit)
}

// cleanupLoop periodically removes stale entries to prevent memory leak
func (rl *RateLimiter) cleanupLoop() {
	ticker := time.NewTicker(1 * time.Minute)
	defer ticker.Stop()

	for {
		select {
		case <-rl.quit:
			return
		case <-ticker.C:
			rl.cleanup()
		}
	}
}

// cleanup removes expired blocks and stale counter entries
func (rl *RateLimiter) cleanup() {
	rl.mu.Lock()
	defer rl.mu.Unlock()

	now := time.Now()
	interval := time.Duration(rl.config.IntervalSeconds) * time.Second
	if interval == 0 {
		interval = 60 * time.Second // Default 1 minute
	}
	// Keep entries for 2x interval for safety margin
	staleThreshold := interval * 2

	// Clean expired blocks
	for ip, expiry := range rl.blocks {
		if now.After(expiry) {
			delete(rl.blocks, ip)
		}
	}

	// Clean stale counters
	for ip, state := range rl.counters {
		if now.Sub(state.lastSeen) > staleThreshold {
			delete(rl.counters, ip)
		}
	}
}

// Check returns true if the IP is allowed, false if blocked/rate-limited
func (rl *RateLimiter) Check(ip string) bool {
	rl.mu.Lock()
	defer rl.mu.Unlock()

	// 1. Check if globally blocked by this limiter
	if expiry, blocked := rl.blocks[ip]; blocked {
		if time.Now().Before(expiry) {
			return false
		}
		// Block expired
		delete(rl.blocks, ip)

		// If we were blocked and are now unblocked, set state so next failure triggers immediate block
		// (fail2ban style "probation")
		if state, exists := rl.counters[ip]; exists && rl.config.CountOnlyFailures {
			// Set to Max - 1, so 1 more failure triggers block
			state.failedCount = int(rl.config.MaxConnections) - 1
			if state.failedCount < 0 {
				state.failedCount = 0
			}
			// Reset window to now, so we have a fresh interval for this probation
			state.firstSeen = time.Now()
		}
	}

	// 2. Check rate limit window
	state, exists := rl.counters[ip]
	now := time.Now()

	if !exists {
		rl.counters[ip] = &ipState{
			count:     0, // Will be incremented in Track
			firstSeen: now,
			lastSeen:  now,
		}
		return true
	}

	// Reset window if interval passed
	if now.Sub(state.firstSeen) > time.Duration(rl.config.IntervalSeconds)*time.Second {
		state.count = 0
		// Only reset failedCount if we haven't just come out of a block (checked above)
		// Actually, if interval passed, we should reset.
		// But if we just unblocked (above), we reset firstSeen to now, so this won't trigger immediately.
		state.failedCount = 0
		state.firstSeen = now
	}

	// If we are NOT counting failures only, check strict connection count
	if !rl.config.CountOnlyFailures {
		if state.count >= int(rl.config.MaxConnections) {
			// Trigger block if auto-block is enabled
			if rl.config.AutoBlock {
				rl.blocks[ip] = now.Add(time.Duration(rl.config.BlockDurationSeconds) * time.Second)
			}
			return false
		}
	}

	return true
}

// TrackConnection records a new connection attempt
func (rl *RateLimiter) TrackConnection(ip string) {
	rl.mu.Lock()
	defer rl.mu.Unlock()

	state, exists := rl.counters[ip]
	if !exists {
		// Should have been created in Check, but just in case
		rl.counters[ip] = &ipState{
			count:     1,
			firstSeen: time.Now(),
		}
		return
	}
	state.count++
	state.lastSeen = time.Now()
}

// ReportResult reports the outcome of a connection (duration, success/fail)
// Used for "fail2ban" logic where we only care about failed attempts
func (rl *RateLimiter) ReportResult(ip string, duration time.Duration) {
	if !rl.config.CountOnlyFailures {
		return
	}

	rl.mu.Lock()
	defer rl.mu.Unlock()

	state, exists := rl.counters[ip]
	if !exists {
		return
	}

	// Define "Failure": Connection duration < Threshold
	threshold := time.Duration(rl.config.FailureDurationThreshold) * time.Second
	if threshold == 0 {
		threshold = 1 * time.Second // Default default
	}

	if duration < threshold {
		state.failedCount++

		// Check limit based on failures
		if state.failedCount >= int(rl.config.MaxConnections) {
			if rl.config.AutoBlock {
				var blockDuration time.Duration

				// Escalation logic
				steps := rl.config.BlockStepsSeconds
				if len(steps) > 0 {
					level := state.banLevel
					if level >= len(steps) {
						level = len(steps) - 1
					}
					blockDuration = time.Duration(steps[level]) * time.Second
					state.banLevel++
				} else {
					// Legacy/Simple mode
					seconds := rl.config.BlockDurationSeconds
					if seconds == 0 {
						seconds = 600 // Default 10 min
					}
					blockDuration = time.Duration(seconds) * time.Second
				}

				rl.blocks[ip] = time.Now().Add(blockDuration)
			}
		}
	}
}

File: nitellad-rs/src/ratelimit.rs
use crate::proto::proxy::RateLimitConfig;
use dashmap::DashMap;
use parking_lot::Mutex;
use std::net::IpAddr;
use std::time::{Duration, Instant};

#[derive(Debug)]
pub struct RateLimiter {
    config: RateLimitConfig,
    counters: DashMap<IpAddr, IpState>,
    blocks: DashMap<IpAddr, Instant>,
    last_cleanup: Mutex<Instant>,
}

#[derive(Debug)]
struct IpState {
    count: i32,
    first_seen: Instant,
    last_seen: Instant,
    failed_count: i32,
    ban_level: usize,
}

impl RateLimiter {
    pub fn new(config: RateLimitConfig) -> Self {
        let now = Instant::now();
        Self {
            config,
            counters: DashMap::new(),
            blocks: DashMap::new(),
            last_cleanup: Mutex::new(now),
        }
    }

    pub fn check(&self, ip: &str) -> bool {
        let Some(ip) = parse_ip(ip) else {
            return true;
        };

        let now = Instant::now();
        self.cleanup_if_due(now);

        let blocked_until = self.blocks.get(&ip).map(|expiry| *expiry);
        if let Some(expiry) = blocked_until {
            if now < expiry {
                return false;
            }
            self.blocks.remove(&ip);
            if let Some(mut counter) = self.counters.get_mut(&ip) {
                counter.failed_count = (self.config.max_connections - 1).max(0);
                counter.first_seen = now;
            }
        }

        let interval = self.interval_duration();
        let block_until = {
            let mut counter = self.counters.entry(ip).or_insert_with(|| IpState::new(now));

            if now.duration_since(counter.first_seen) > interval {
                counter.count = 0;
                counter.failed_count = 0;
                counter.first_seen = now;
            }
            counter.last_seen = now;

            if !self.config.count_only_failures && counter.count >= self.config.max_connections {
                if self.config.auto_block {
                    Some(now + self.block_duration(counter.ban_level))
                } else {
                    None
                }
            } else {
                return true;
            }
        };

        if let Some(expiry) = block_until {
            self.blocks.insert(ip, expiry);
        }
        false
    }

    pub fn track_connection(&self, ip: &str) {
        let Some(ip) = parse_ip(ip) else {
            return;
        };

        let now = Instant::now();
        let mut counter = self.counters.entry(ip).or_insert_with(|| IpState::new(now));
        counter.count += 1;
        counter.last_seen = now;
    }

    pub fn report_result(&self, ip: &str, duration: Duration) {
        if !self.config.count_only_failures {
            return;
        }

        let threshold = if self.config.failure_duration_threshold > 0 {
            Duration::from_secs(self.config.failure_duration_threshold as u64)
        } else {
            Duration::from_secs(1)
        };
        if duration >= threshold {
            return;
        }

        let Some(ip) = parse_ip(ip) else {
            return;
        };

        let now = Instant::now();

        let block_until = {
            let mut counter = match self.counters.get_mut(&ip) {
                Some(counter) => counter,
                None => return,
            };
            counter.failed_count += 1;
            counter.last_seen = now;

            if counter.failed_count >= self.config.max_connections && self.config.auto_block {
                let block_duration = self.block_duration(counter.ban_level);
                counter.ban_level += 1;
                Some(now + block_duration)
            } else {
                None
            }
        };

        if let Some(expiry) = block_until {
            self.blocks.insert(ip, expiry);
        }
    }

    fn cleanup_if_due(&self, now: Instant) {
        let interval = self.interval_duration();
        {
            let mut last_cleanup = self.last_cleanup.lock();
            if now.duration_since(*last_cleanup) < interval {
                return;
            }
            *last_cleanup = now;
        }

        let stale_threshold = self.interval_duration() * 2;
        self.blocks.retain(|_, expiry| now < *expiry);
        self.counters
            .retain(|_, counter| now.duration_since(counter.last_seen) <= stale_threshold);
    }

    fn interval_duration(&self) -> Duration {
        if self.config.interval_seconds > 0 {
            Duration::from_secs(self.config.interval_seconds as u64)
        } else {
            Duration::from_secs(60)
        }
    }

    fn block_duration(&self, ban_level: usize) -> Duration {
        if !self.config.block_steps_seconds.is_empty() {
            let idx = ban_level.min(self.config.block_steps_seconds.len() - 1);
            return Duration::from_secs(self.config.block_steps_seconds[idx].max(0) as u64);
        }
        if self.config.block_duration_seconds > 0 {
            Duration::from_secs(self.config.block_duration_seconds as u64)
        } else {
            Duration::from_secs(600)
        }
    }
}

impl IpState {
    fn new(now: Instant) -> Self {
        Self {
            count: 0,
            first_seen: now,
            last_seen: now,
            failed_count: 0,
            ban_level: 0,
        }
    }
}

fn parse_ip(ip: &str) -> Option<IpAddr> {
    ip.parse().ok()
}

#[cfg(test)]
mod tests {
    use super::*;

    fn config(max_connections: i32) -> RateLimitConfig {
        RateLimitConfig {
            max_connections,
            interval_seconds: 60,
            auto_block: true,
            block_duration_seconds: 60,
            ..Default::default()
        }
    }

    #[test]
    fn blocks_after_max_connections_in_window() {
        let limiter = RateLimiter::new(config(1));
        assert!(limiter.check("10.0.0.1"));
        limiter.track_connection("10.0.0.1");
        assert!(!limiter.check("10.0.0.1"));
    }

    #[test]
    fn count_only_failures_blocks_after_short_results() {
        let limiter = RateLimiter::new(RateLimitConfig {
            count_only_failures: true,
            failure_duration_threshold: 10,
            ..config(1)
        });
        assert!(limiter.check("10.0.0.2"));
        limiter.track_connection("10.0.0.2");
        limiter.report_result("10.0.0.2", Duration::from_secs(1));
        assert!(!limiter.check("10.0.0.2"));
    }
}
